diff --git CMakeLists.txt CMakeLists.txt
index 43c73e3e..0e910d11 100644
--- CMakeLists.txt
+++ CMakeLists.txt
@@ -121,7 +121,7 @@ find_package(ESMF 8.3.0 MODULE REQUIRED)
 if(FMS)
   find_package(FMS 2022.01 REQUIRED COMPONENTS R4 R8)
   if(32BIT)
-    add_library(fms ALIAS FMS::fms_r4)
+    add_library(fms ALIAS FMS::fms_r8)
   else()
     add_library(fms ALIAS FMS::fms_r8)
   endif()
diff --git modulefiles/ufs_common modulefiles/ufs_common
index 440ba826..ee291a86 100644
--- modulefiles/ufs_common
+++ modulefiles/ufs_common
@@ -8,7 +8,7 @@ module load hdf5/1.10.6
 module load netcdf/4.7.4
 module load pio/2.5.3
 module load esmf/8.3.0b09
-module load fms/2022.01
+module load fms/2022.03

 module load bacio/2.4.1
 module load crtm/2.3.0
