--- MOM6-interface/MOM6/config_src/drivers/nuopc_cap/mom_cap.F90
+++ MOM6-interface/MOM6/config_src/drivers/nuopc_cap/mom_cap.F90
@@ -1525,6 +1525,8 @@ subroutine ModelAdvance(gcomp, rc)
   integer, intent(out)                   :: rc    !< return code
 
   ! local variables
+  logical                                :: restart_endfcst = .false.
+  logical                                :: write_restart = .true.
   integer                                :: userRc
   logical                                :: existflag, isPresent, isSet
   logical                                :: do_advance = .true.
@@ -1765,6 +1767,9 @@ subroutine ModelAdvance(gcomp, rc)
       else  ! not cesm_coupled
         ! write the final restart without a timestamp
         if (ESMF_AlarmIsRinging(stop_alarm, rc=rc)) then
+          if (.not. restart_endfcst) then
+             write_restart = .false.
+          endif 
           write(restartname,'(A)')"MOM.res"
           write(stoch_restartname,'(A)')"ocn_stoch.res.nc"
         else
@@ -1776,8 +1781,10 @@ subroutine ModelAdvance(gcomp, rc)
         call ESMF_LogWrite("MOM_cap: Writing restart :  "//trim(restartname), ESMF_LOGMSG_INFO)
 
         ! write restart file(s)
-        call ocean_model_restart(ocean_state, restartname=restartname, &
-                                stoch_restartname=stoch_restartname)
+        if (write_restart) then
+            call ocean_model_restart(ocean_state, restartname=restartname, &
+                                    stoch_restartname=stoch_restartname)
+        endif
 
       endif
 
